diff -ruN busybox-1.37.0/init/init.c busybox-1.37.0-patch/init/init.c
--- busybox-1.37.0/init/init.c	2024-04-16 10:23:50.000000000 +0300
+++ busybox-1.37.0-patch/init/init.c	2025-01-10 14:43:56.377585206 +0300
@@ -813,6 +813,8 @@
 	const char *m;
 	unsigned rb;
 
+	const char *bbs;
+
 	/* We may call run() and it unmasks signals,
 	 * including the one masked inside this signal handler.
 	 * Testcase which would start multiple reboot scripts:
@@ -821,6 +823,10 @@
 	 */
 	reset_sighandlers_and_unblock_sigs();
 
+	bbs = "shutdown";
+	if (sig == SIGTERM) bbs = "reboot";
+
+	xsetenv("BB_SHUTDOWN", bbs);
 	run_shutdown_and_kill_processes();
 
 	m = "halt";
